File: scripts/world.py
import sys
from array import array
import json
import math

if len(sys.argv) != 3:
    print(f"ERROR: Syntax: py {sys.argv[0]} input_file output_file")
    exit()

inputfilename = sys.argv[1]
outputfilename = sys.argv[2]

inputfile = open(inputfilename, "r")
outputfile = open(outputfilename, "wb")
data = json.loads(inputfile.read())

outputfile.write(array("b", [0x53, 0x57, 0, 0]))
outputfile.write(array("i", [len(data["models"])]))
outputfile.write(array("i", [len(data["textures"])]))
outputfile.write(array("i", [len(data["objects"])]))
outputfile.write(array("i", [len(data["point-cuboids"])]))

models = []
textures = []

for index, model in enumerate(data["models"]):
    name = model["name"]
    file = model["file"]

    zero_count = 20 - len(file)

    outputfile.write(array("b", file.encode()))
    outputfile.write(array("b", [0] * zero_count))

    models.append(name)

for index, texture in enumerate(data["textures"]):
    name = texture["name"]
    file = texture["file"]

    zero_count = 20 - len(file)

    outputfile.write(array("b", file.encode()))
    outputfile.write(array("b", [0] * zero_count))

    textures.append(name)

for object in data["objects"]:
    model_index = models.index(object["model"])
    texture_index = textures.index(object["texture"])
    position = object.get("position", [0, 0, 0])
    rotation = object.get("rotation", [0, 0, 0])
    scale = object.get("scale", [1, 1, 1])

    position[1], position[2] = -position[2], position[1]  # Vulkan
    rotation[1], rotation[2] = -rotation[2], rotation[1]  # Vulkan
    scale[1], scale[2] = -scale[2], scale[1]  # Vulkan

    rotation = [math.radians(x) for x in rotation]

    outputfile.write(array("i", [model_index, texture_index]))
    outputfile.write(array("f", position))
    outputfile.write(array("f", rotation))
    outputfile.write(array("f", scale))

for cb in data["point-cuboids"]:
    centre = cb.get("centre", [0, 0, 0])
    centre[1], centre[2] = -centre[2], centre[1]  # Vulkan
    dimension = cb.get("dimension", [0, 0, 0])
    dimension[1], dimension[2] = dimension[2], dimension[1]  # Vulkan

    outputfile.write(array("f", centre))
    outputfile.write(array("f", dimension))

File: src/world/world.c
/**
 * @brief Load a .sm file from disk and create buffers for Vulkan
 * 
 * @param device Vulkan device
 * @param physical_device Vulkan physical device
 * @param filename Name of the file to read and parse
 * @return struct sln_world Information about the created world
 */
struct sln_world sln_load_sw(
    VkDevice device,
    VkPhysicalDevice physical_device,
    VkCommandPool command_pool,
    VkQueue graphics_queue,
    VkSampler sampler,
    VkDescriptorPool pool,
    VkDescriptorSetLayout set_layout,
    char *filename
){
    struct sln_world world = {0};
    struct sln_file file = sln_read_file(filename, 1);
    struct sw_header *header = (struct sw_header *)file.data;

    // Models
    struct sw_model *model_ptr = (struct sw_model *)((uint64_t)header +
        sizeof(struct sw_header));

    for (uint32_t i = 0; i < header->model_count; i++, model_ptr++)
        world.models[i] = sln_load_sm(device, physical_device,
            model_ptr->filename);

    // Textures
    struct sw_model *texture_ptr = (struct sw_model *)model_ptr;

    for (uint32_t i = 0; i < header->texture_count; i++, texture_ptr++)
        world.textures[i] = sln_load_simg(device, physical_device, command_pool,
            graphics_queue, sampler, pool, set_layout, texture_ptr->filename);

    // Objects
    struct sw_object *object_ptr = (struct sw_object *)texture_ptr;

    for (uint32_t i = 0; i < header->object_count; i++, object_ptr++) {
        struct sln_object *object = &world.objects[i];
        object->model_index = object_ptr->model_index;
        object->texture_index = object_ptr->texture_index;
        object->flags = SLN_WORLD_FLAG_EXISTS;
        tf_set(&object->transform, object_ptr->position, object_ptr->rotation,
            object_ptr->scale);
    }

    // Point Cuboids
    struct sw_point_cuboid *cb_ptr = (struct sw_point_cuboid *)object_ptr;

    for (uint32_t i = 0; i < header->point_cuboid_count; i++, cb_ptr++) {
        struct point_cuboid *cb = &world.physics.cuboid[i];
        cb->flags = PHYSICS_FLAG_EXISTS;
        cb->centre.x = cb_ptr->centre[0];
        cb->centre.y = cb_ptr->centre[1];
        cb->centre.z = cb_ptr->centre[2];

        cb->dimension.x = cb_ptr->dimension[0];
        cb->dimension.y = cb_ptr->dimension[1];
        cb->dimension.z = cb_ptr->dimension[2];
    }

    sln_close_file(file);
    return world;
}
